File: src/normalize.ts
const APOSTROPHE_LIKE = new Set(['ʻ', 'ʼ', "'", '‘', '’', '`', '´', '′', '＇']);
const LETTER_MODIFIER_PREDECESSORS = new Set(['o', 'O', 'g', 'G']);
const CONFUSABLES: Readonly<Record<string, string>> = { ș: 'ş', Ș: 'Ş' };

export function normalizeConfusables(text: string): string {
  let result = '';
  for (const character of text.normalize('NFC')) result += CONFUSABLES[character] ?? character;
  return result.normalize('NFC');
}

export function normalizeApostrophes(text: string): string {
  const source = normalizeConfusables(text);
  let result = '';
  let previous = '';
  for (const character of source) {
    const normalized = APOSTROPHE_LIKE.has(character)
      ? LETTER_MODIFIER_PREDECESSORS.has(previous)
        ? 'ʻ'
        : 'ʼ'
      : character;
    result += normalized;
    previous = character;
  }
  return result.normalize('NFC');
}

export function prepareText(text: string): string {
  return normalizeApostrophes(normalizeConfusables(text.normalize('NFC')));
}

File: python/src/alifbo/normalize.py
"""NFC, confusable, and apostrophe normalization (mirrors ``src/normalize.ts``)."""

from __future__ import annotations

import unicodedata

_APOSTROPHE_LIKE = frozenset(
    (
        "ʻ",  # MODIFIER LETTER TURNED COMMA
        "ʼ",  # MODIFIER LETTER APOSTROPHE
        "'",
        "‘",
        "’",
        "`",
        "´",
        "′",
        "＇",
    )
)
_LETTER_MODIFIER_PREDECESSORS = frozenset(("o", "O", "g", "G"))
_CONFUSABLES = {"ș": "ş", "Ș": "Ş"}


def nfc(text: str) -> str:
    return unicodedata.normalize("NFC", text)


def normalize_confusables(text: str) -> str:
    """NFC-normalize text and fold known visual confusables to their canonical Uzbek forms."""
    return nfc("".join(_CONFUSABLES.get(character, character) for character in nfc(text)))


def normalize_apostrophes(text: str) -> str:
    """Collapse apostrophe-like characters to U+02BB after o/g and U+02BC elsewhere."""
    source = normalize_confusables(text)
    result = []
    previous = ""
    for character in source:
        if character in _APOSTROPHE_LIKE:
            result.append("ʻ" if previous in _LETTER_MODIFIER_PREDECESSORS else "ʼ")
        else:
            result.append(character)
        previous = character
    return nfc("".join(result))


def prepare_text(text: str) -> str:
    return normalize_apostrophes(normalize_confusables(nfc(text)))
